src/modules/ekf2/EKF/ekf_helper.cpp

```diff
@@ -321,39 +322,43 @@ bool Ekf::setEkfGlobalOrigin(const double latitude, const double longitude, cons
 
 			_wmm_gps_time_last_set = _time_delayed_us;
 		}
+
 #endif // CONFIG_EKF2_MAGNETOMETER
 
 		_gpos_origin_eph = eph;
 		_gpos_origin_epv = epv;
 
 		_NED_origin_initialised = true;
 
-		// minimum change in position or height that triggers a reset
-		static constexpr float MIN_RESET_DIST_M = 0.01f;
-
 		if (current_pos_available) {
-			// reset horizontal position
+			// reset horizontal position if we already have a global origin
 			Vector2f position = _pos_ref.project(current_lat, current_lon);
-
-			if (Vector2f(position - Vector2f(_state.pos)).longerThan(MIN_RESET_DIST_M)) {
-				resetHorizontalPositionTo(position);
-			}
+			resetHorizontalPositionTo(position);
 		}
 
-		// reset vertical position (if there's any change)
-		if (fabsf(altitude - gps_alt_ref_prev) > MIN_RESET_DIST_M) {
-			// determine current z
-			float current_alt = -_state.pos(2) + gps_alt_ref_prev;
+		if (PX4_ISFINITE(gps_alt_ref_prev) && isVerticalPositionAidingActive()) {
+			// reset vertical position if we already had a GPS altitude reference
 
-#if defined(CONFIG_EKF2_GNSS)
-			const float gps_hgt_bias = _gps_hgt_b_est.getBias();
-#endif // CONFIG_EKF2_GNSS
-			resetVerticalPositionTo(_gps_alt_ref - current_alt);
+			ECL_DEBUG("EKF global origin updated, GPS altitude ref %.1fm -> %.1fm", (double)gps_alt_ref_prev, (double)_gps_alt_ref);
+
+			if (_height_sensor_ref == HeightSensor::GNSS) {
+				// determine current z
+				float z_prev = _state.pos(2);
+				float current_alt = -z_prev + gps_alt_ref_prev;
+				resetVerticalPositionTo(_gps_alt_ref - current_alt);
+
+				ECL_DEBUG("EKF global origin updated, resetting vertical position %.1fm -> %.1fm", (double)z_prev, (double)_state.pos(2));
 
+			} else {
 #if defined(CONFIG_EKF2_GNSS)
-			// preserve GPS height bias
-			_gps_hgt_b_est.setBias(gps_hgt_bias);
+				// adjust existing GPS height bias
```

Comment:
@dagar I don't quite follow the logic. How come you don't reset the vertical position if the height reference is not GNSS? Since we change the origin altitude, would that not lead to a wrong global altitude?